src/Microsoft.AspNet.OData.Shared/Query/ExpandQueryBuilder.cs

```diff
@@ -0,0 +1,121 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="ExpandQueryBuilder.cs" company=".NET Foundation">
+//      Copyright (c) .NET Foundation and Contributors. All rights reserved. 
+//      See License.txt in the project root for license information.
+// </copyright>
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.AspNet.OData.Formatter;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    /// <summary>
+    /// Exposes the ability to generate a $expand query string from a payload object.
+    /// </summary>
+    public class ExpandQueryBuilder : IExpandQueryBuilder
+    {
+        /// <inheritdoc />
+        public virtual string GenerateExpandQueryString(object value, IEdmModel model)
+        {
+            if (value == null)
+            {
+                throw Error.ArgumentNull(nameof(value));
+            }
+
+            if (model == null)
+            {
+                throw Error.ArgumentNull(nameof(model));
+            }
+
+            return GenerateExpandQueryStringInternal(value, model, true);
+        }
+
+        private string GenerateExpandQueryStringInternal(object value, IEdmModel model, bool isFirstExpand)
+        {
+            Type type = value.GetType();
+            Type _elementType = null;
+            bool isCollection = TypeHelper.IsCollection(type, out _elementType);
+            IList objList = null;
+            if (isCollection)
+            {
+                type = _elementType;
+                objList = value as IList;
+            }
+
+            string edmFullName = type.EdmFullName();
+            IEdmSchemaType schemaType = model.FindType(edmFullName);
+            IEdmStructuredType edmStructuredType = schemaType as IEdmStructuredType;
+
+            IEnumerable<IEdmNavigationProperty> navigationProperties = edmStructuredType.NavigationProperties();
+            string expandString = "";
+
+            int count = 0;
+
+            List<string> navPropNames = new List<string>();
+
+            foreach (IEdmNavigationProperty navProp in navigationProperties)
+            {
+                count++;
+                PropertyInfo prop = type.GetProperty(navProp.Name);
+
+                if (isCollection)
+                {
+                    foreach (object obj in objList)
+                    {
+                        object listItem = prop.GetValue(obj);
+
+                        if (listItem != null)
+                        {
+                            if (!navPropNames.Contains(navProp.Name))
```

Comment:
What if I have the following kind of structure:
```csharp
public class Person
{
    public int Id { get; set; }
    public Person BestFriend { get; set; }
}
```
and I pass the following object into `GenerateExpandQueryString`:
```csharp
new Person
{
    Id = 7,
    BestFriend = new Person
    {
        Id = 13,
        BestFriend = new Person
        {
            Id = 17
        }
    }
}
```
What I would be intending here is to get back:
```
$expand=BestFriend($expand=BestFriend)
```
But because of the `!navPropNames.Contains(navProp.Name)` check, I'd get back:
```
$expand=BestFriend
```

Like I observed in a previous comment, you probably don't need to loop through the collection since the items are homogeneous. And if they were not homogeneous, e.g. collection contains `Order` and `VipOrder` where `VipOrder` could have navigation properties that `Order` doesn't have, your logic here still wouldn't handle it smoothly

Reply:
@KenitoInc On second thoughts, I think this scenario would be handled okay since the method would be called recursively and `navPropNames` would be reintialized... But it wouldn't hurt to add a test for such a scenario.

Reply:
Exactly. It's recursion so `navPropNames` is populated for every level.